geopandas/tools/overlay.py

```diff
@@ -331,69 +340,69 @@ def _make_valid(df):
-            geom_col = result._geometry_column_name
-            collections = result[[geom_col]][is_collection]
-
-            exploded = collections.reset_index(drop=True).explode(index_parts=True)
-            exploded = exploded.reset_index(level=0)
-
-            orig_num_geoms_exploded = exploded.shape[0]
-            if geom_type in polys:
-                exploded.loc[~exploded.geom_type.isin(polys), geom_col] = None
-            elif geom_type in lines:
-                exploded.loc[~exploded.geom_type.isin(lines), geom_col] = None
-            elif geom_type in points:
-                exploded.loc[~exploded.geom_type.isin(points), geom_col] = None
-            else:
-                raise TypeError(
-                    "`keep_geom_type` does not support {}.".format(geom_type)
-                )
-            num_dropped_collection = (
-                orig_num_geoms_exploded - exploded.geometry.isna().sum()
-            )
+    result.reset_index(drop=True, inplace=True)
+    return result
 
-            # level_0 created with above reset_index operation
-            # and represents the original geometry collections
-            # TODO avoiding dissolve to call union_all in this case could further
-            # improve performance (we only need to collect geometries in their
-            # respective Multi version)
-            dissolved = exploded.dissolve(by="level_0")
-            result.loc[is_collection, geom_col] = dissolved[geom_col].values
-        else:
-            num_dropped_collection = 0
-
-        # Now we filter all geometries (in theory we don't need to do this
-        # again for the rows handled above for GeometryCollections, but filtering
-        # them out is probably more expensive as simply including them when this
-        # is typically about only a few rows)
-        orig_num_geoms = result.shape[0]
-        if geom_type in polys:
-            result = result.loc[result.geom_type.isin(polys)]
-        elif geom_type in lines:
-            result = result.loc[result.geom_type.isin(lines)]
-        elif geom_type in points:
-            result = result.loc[result.geom_type.isin(points)]
-        else:
-            raise TypeError("`keep_geom_type` does not support {}.".format(geom_type))
-        num_dropped = orig_num_geoms - result.shape[0]
-
-        if (num_dropped > 0 or num_dropped_collection > 0) and keep_geom_type_warning:
-            warnings.warn(
-                "`keep_geom_type=True` in overlay resulted in {} dropped "
-                "geometries of different geometry types than df1 has. "
-                "Set `keep_geom_type=False` to retain all "
-                "geometries".format(num_dropped + num_dropped_collection),
-                UserWarning,
-                stacklevel=2,
-            )
 
-    result.reset_index(drop=True, inplace=True)
+def _collection_extract(df, geom_type, keep_geom_type_warning):
+    # Check input
+    if geom_type in ["Polygon", "MultiPolygon"]:
```

Comment:
Would it make sense to define e.g. `POLYGON_GEOM_TYPES = ["Polygon", "MultiPolygon"]` as constants now these are used across a couple of places?

This is a nice simplification of the original logic, good to see it split out into a seperate method.

Reply:
OK, I created constants for the different "subtypes"... I put them in array.py because there were already similar constants there, not sure if that is the ideal spot...

Reply:
> This is a nice simplification of the original logic, good to see it split out into a seperate method.

Ideally a more generic version would become available in shapely... https://github.com/shapely/shapely/issues/1927, but step by step...

Reply:
Yes indeed, incremental progress is still progress.